skada/deep/_adversarial.py

```diff
@@ -370,6 +369,148 @@ def CDAN(
+        self.gamma = gamma
+
+    def forward(
+        self,
+        y_s,
+        y_pred_s,
+        y_pred_t,
+        domain_pred_s,
+        domain_pred_t,
+        features_s,
+        features_t,
+    ):
+        """Compute the domain adaptation loss"""
+        # TODO: handle binary classification
+        # Multiclass classification
+        pseudo_label_s = torch.argmax(y_pred_s, axis=-1)
+        pseudo_label_t = torch.argmax(y_pred_t, axis=-1)
+
+        disc_loss_src = self.domain_criterion_(domain_pred_s, pseudo_label_s)
+        disc_loss_tgt = self.domain_criterion_(domain_pred_t, pseudo_label_t)
+
+        # Compute the MDD loss value
+        disc_loss = self.gamma * disc_loss_src - disc_loss_tgt
+
+        return disc_loss
+
+
+def MDD(
+    module,
+    layer_name,
+    reg=1,
+    gamma=4.0,
+    domain_classifier=None,
+    num_features=None,
+    n_classes=None,
+    base_criterion=None,
+    domain_criterion=None,
+    **kwargs,
+):
+    """Margin Disparity Discrepancy (MDD).
+
+    From [35]_.
+
+    Parameters
+    ----------
+    module : torch module (class or instance)
+        A PyTorch :class:`~torch.nn.Module`. In general, the
+        uninstantiated class should be passed, although instantiated
+        modules will also work.
+    layer_name : str
+        The name of the module's layer whose outputs are
+        collected during the training.
+    reg : float, default=1
+        Regularization parameter for DA loss.
+    domain_classifier : torch module, default=None
+        A PyTorch :class:`~torch.nn.Module` used to classify the
+        domain. If None, a domain classifier is created following [1]_.
+    num_features : int, default=None
+        Size of the input of domain classifier,
+        e.g size of the last layer of
+        the feature extractor.
+        If domain_classifier is None, num_features has to be
+        provided.
+    n_classes : int, default=None
+        Number of classes. If domain_classifier is None,
+        n_classes has to be provided.
+    base_criterion : torch criterion (class)
+        The base criterion used to compute the loss with source
+        labels. If None, the default is `torch.nn.CrossEntropyLoss`.
+    domain_criterion : torch criterion (class)
+        The criterion (loss) used to compute the
+        MDD loss. If None, a CrossEntropyLoss is used.
+    gamma : float (default=4.0)
+        Margin parameter following [35]_.
+
+    References
+    ----------
+    .. [35] Yuchen Zhang et. al. Bridging Theory and Algorithm
+            for Domain Adaptation. In International Conference on
+            Machine Learning, 2019.
+    """
+    if domain_classifier is None:
+        # raise error if num_feature is None
+        if num_features is None:
+            raise ValueError(
+                "If domain_classifier is None, num_features has to be provided"
+            )
+        domain_classifier = DomainClassifier(
```

Comment:
I don't know if we use here the domain Classifier or if we want a `DiscClassifier` maybe the same that is in the paper

Reply:
In the paper, the disc_classifier must be the same architecture as the task classifier. So in our case, the same than the one used with base_criterion